packages/twenty-server/src/engine/metadata-modules/ai-router/services/ai-router-plan-generator.service.ts

```diff
@@ -0,0 +1,158 @@
+import { Injectable, Logger } from '@nestjs/common';
+
+import {
+  generateObject,
+  type UIDataTypes,
+  type UIMessage,
+  type UITools,
+} from 'ai';
+import { z } from 'zod';
+
+import { type ModelId } from 'src/engine/core-modules/ai/constants/ai-models.const';
+import { AI_TELEMETRY_CONFIG } from 'src/engine/core-modules/ai/constants/ai-telemetry.const';
+import { AiModelRegistryService } from 'src/engine/core-modules/ai/services/ai-model-registry.service';
+import { type AgentEntity } from 'src/engine/metadata-modules/agent/agent.entity';
+import { AGENT_SYSTEM_PROMPTS } from 'src/engine/metadata-modules/agent/constants/agent-system-prompts.const';
+import { type ExecutionPlan } from 'src/engine/metadata-modules/ai-router/types/router-result.interface';
+
+@Injectable()
+export class AiRouterPlanGeneratorService {
+  private readonly logger = new Logger(AiRouterPlanGeneratorService.name);
+
+  constructor(
+    private readonly aiModelRegistryService: AiModelRegistryService,
+  ) {}
+
+  async generatePlan({
+    messages,
+    availableAgents,
+    agentDescriptions,
+    plannerModel,
+  }: {
+    messages: UIMessage<unknown, UIDataTypes, UITools>[];
+    availableAgents: AgentEntity[];
+    agentDescriptions: string;
+    plannerModel: ModelId;
+  }): Promise<ExecutionPlan> {
+    const model = this.getPlannerModel(plannerModel);
+    const agentNames = availableAgents.map((agent) => agent.name);
+
+    const conversationHistory = messages
```

Comment:
not sure we can do better but this code looks very bug prone (array manipulation)